src/celeritas/global/alongstep/AlongStep.hh

```diff
@@ -0,0 +1,134 @@
+#include "celeritas/Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Within-kernel state for step limits and action
+struct AlongStepLocalState
+{
+    //! Final step limit (true path length) and action
+    StepLimit step_limit;
+    //! Smooth (straight/curved) step movement without accounting for MSC
+    real_type geo_step{};
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform the along-step action using helper functions.
+ *
+ * \tparam MH MSC helper, e.g. \c detail::NoMsc
+ * \tparam MP Propagator factory, e.g. \c detail::LinearPropagatorFactory
+ * \tparam AE Energy loss applier, e.g. \c detail::NoElossApplier
+ */
+template<class MH, class MP, class AE>
+inline CELER_FUNCTION void along_step(MH&&                 msc,
+                                      MP&&                 make_propagator,
+                                      AE&&                 apply_eloss,
+                                      CoreTrackView const& track)
+{
+    // TODO: scope the 'views' so that the lifetimes don't overlap between this
+    // function and helper class functions
+    auto sim = track.make_sim_view();
+
+    // True step is the actual path length traveled by the particle, including
+    // within-step MSC
+    AlongStepLocalState local;
+    local.step_limit = sim.step_limit();
+    CELER_ASSERT(local.step_limit);
+    if (local.step_limit.step == 0)
+    {
+        // Track is stopped: no movement or energy loss will happen
+        // (could be a stopped positron waiting for annihilation, or a particle
+        // waiting to decay?)
+        CELER_ASSERT(track.make_particle_view().is_stopped());
+        CELER_ASSERT(local.step_limit.action
+                     == track.make_physics_view().scalars().discrete_action());
+        // Increment the step counter
+        sim.increment_num_steps();
+        return;
+    }
+
+    local.geo_step = local.step_limit.step;
+    bool use_msc   = msc.is_applicable(track, local.geo_step);
+    if (use_msc)
+    {
+        msc.calc_step(track, &local);
+        CELER_ASSERT(local.geo_step > 0);
+        CELER_ASSERT(local.step_limit.step >= local.geo_step);
+    }
+
+    {
+        auto geo       = track.make_geo_view();
+        auto propagate = make_propagator(track.make_particle_view(), &geo);
+        Propagation p  = propagate(local.geo_step);
+        if (p.boundary)
+        {
+            // Stopped at a geometry boundary: this is the new step action.
+            local.geo_step          = p.distance;
+            local.step_limit.action = track.boundary_action();
+        }
+    }
+
+    if (use_msc)
+    {
+        msc.apply_step(track, &local);
+    }
+    else
+    {
+        // Step might have changed due to geometry boundary
+        local.step_limit.step = local.geo_step;
+    }
+
+    // Update track's lab-frame time using the beginning-of-step speed
```

Comment:
Wondering whether this block should go before msc.apply_step (i.e, time update based on the geometry step, not by true step - I am not sure which one is correct, but Geant4, as far as I know, updates the global time before the msc sampling)? In general, the field integration may be performed by the time interval (instead of the distance traveled) alternatively.

Reply:
Section 3.2.4 of the PRM says the time change is calculated using the true path length -- but it also says it's calculated from both the initial and final particle velocity, and as far as I could tell in the code only the initial velocity is used, so it might be good to double check. I think global time is updated [here](https://geant4.kek.jp/lxr/source/tracking/src/G4SteppingManager2.cc#L470), and msc is sampled just before in `AlongStepDoIt`?

Reply:
The sequence of Geant4 AlongStepDoIt goes as Transporation -> msc -> eIoni and the actual place for calculating time in Geant4 is inside G4Transporation::AlongStepDoIt (see this [line](https://geant4.kek.jp/lxr/source/processes/transportation/src/G4Transportation.cc#L529) and update time information based on that).  So, it seems that there is, at least, inconsistency between PRM and the actual time used in the code which is calculated with geometry_step/initialVelocity - in this case, the average time is equivalent to the initial time as there is no change in energy (if there is no electric field).  To me, if the average velocity and the true path should be used, then the time update should be done even after eloni as the energy (so momentum) is changing during the eloss, but not during msc (even though b.

Reply:
ahh I see, thanks @whokion -- good catch!

Reply:
It seems to me we should be using the true path for the update, and I would imagine the average speed would be more accurate. Again, this implementation is equivalent to what's on master, so I'm adding a note on #459 and leaving it as-is for now.